Add neato dismiss alert button

maxcanna opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Adding a _Dismiss Alert_ button to `neato` integration to be able to dismiss alerts preventing cleaning from start

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade https://github.com/stianaske/pybotvac/releases/tag/v0.0.24
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28395

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

maxcanna committed:
Bump pybotvac

gjohansson-ST commented:
Looks like this should be a button?

gjohansson-ST commented:
You can push a button in an automation. 
It's still a service `button.press`

maxcanna commented:
> You can push a button in an automation. It's still a service `button.press`

Yeah, I know but that would be overhead and imho not conceptually correct. It's a service which exposes a functionality

gjohansson-ST commented:
> Yeah, I know but that would be overhead and imho not conceptually correct. It's a service which exposes a functionality 

Can't understand why it would be conceptually wrong? As probably there is a button or something you can click in the app to do this. 
Should still be a button entity.

maxcanna commented:
> As probably there is a button or something you can click in the app to do this.

There's a button in the app to start the cleaning too. Nonetheless the custom cleaning is a service and not a button. I think it's better if it's a service, even for coherence.

I think it's best not to think UI first, and then interact with UI entities in code, but to think the other way around and expose something in the UI only when/if needed.

Rudd-O reviewed (approved):
Oh man this would be great.

joostlek commented on homeassistant/components/neato/services.yaml:
```diff
@@ -30,3 +30,10 @@ custom_cleaning:
       selector:
         text:
+dismiss_alert:
+  name: Dismiss Alert
+  description: Dismiss alert. E.g. dust bin full.
```
Name and description should be in strings.json since 2023.8. Please check other integration with services for examples

maxcanna replied:
oh, right, now services are translatable! 😅  TY, I fixed it

joostlek commented:
I just read the discussion and I share the opinion of @gjohansson-ST. The other service you mentioned takes arguments, so it requires more information to execute that action. In this case, it's something you can fire without extra information.

For automations, making it a service or making it a button makes no difference as both can be called and both would work. But the benefit for the button is in the usability for people who don't use it in an automation. If the button shares the same device info as the vacuum, the button will show up at the device page, allowing users to press it directly from the UI without having to create an integration or to run developer tools to call the service directly.

I found an example on my environment with Plex:
![image](https://github.com/home-assistant/core/assets/7083755/60ce8e87-660d-49da-8684-b93e475a95ce)

gjohansson-ST reviewed (changes requested):
Should be a `button` and not a custom service.
Please change this as already earlier said by myself and another developer.

maxcanna committed:
Add neato dismiss alert button

maxcanna commented:
@gjohansson-ST @joostlek I converted this to a button. I updated the PR title accordingly. 
https://github.com/home-assistant/home-assistant.io/pull/28395 is not needed anymore right? I don't think we report available buttons in the doc. Correct if I'm wrong (telling how to report the new button in there)

gjohansson-ST commented:
> @gjohansson-ST @joostlek I converted this to a button. I updated the PR title accordingly. [home-assistant/home-assistant.io#28395](https://github.com/home-assistant/home-assistant.io/pull/28395) is not needed anymore right? I don't think we report available buttons in the doc. Correct if I'm wrong (telling how to report the new button in there)

At least for the doc's PR you need to update the platform and possibly the category in the header section.
But it might not be a bad thing adding some text what this button does or even an automation example if you can see such case.

gjohansson-ST commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,70 @@
+"""Support for Neato buttons."""
```
Needs test or add to `.coveragerc` to exclude

maxcanna replied:
Ok, I added the newly added file to the exclude

maxcanna commented:
@gjohansson-ST I fixed the tests and updated the doc PR. lmk

joostlek commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,70 @@
+        """Return unique ID."""
+        return self._robot_serial
+
+    @property
+    def entity_category(self) -> EntityCategory:
+        """Device entity category."""
+        return EntityCategory.CONFIG
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Device info for neato robot."""
+        return DeviceInfo(identifiers={(NEATO_DOMAIN, self._robot_serial)})
```
These two can be set with _attr_entity_category and _attr_device_info as well

joostlek commented:
Oh and please include a changelog for the dependency you are updating in the PR description

maxcanna commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,70 @@
+        """Press the button."""
+        await self.hass.async_add_executor_job(self.robot.dismiss_current_alert)
+
+    @property
+    def name(self) -> str:
+        """Return the name of this camera."""
+        return self._button_name
```
Yeah, I know, I kept it this way just to be coherent with existing code (like https://github.com/maxcanna/core/blob/04e193372868d0c841845b68e4648a36a99725d4/homeassistant/components/neato/camera.py#L118). I'll change it asap 😉

maxcanna commented:
> Oh and please include a changelog for the dependency you are updating in the PR description

I linked the release in the PR description

maxcanna committed:
fixup! Add neato dismiss alert button

joostlek commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,52 @@
+        """Initialize a dismiss_alert Neato button entity."""
+        self.robot = robot
+        self._attr_name = f"{self.robot.name} Dismiss Alert"
+        self._attr_unique_id = self.robot.serial
```
The vacuum already uses this unique id, so it's a good practice to add something unique for this entity
```suggestion
        self._attr_unique_id = f"{self.robot.serial}_dismiss_alert"
```

maxcanna replied:
Yes, I noticed that but I saw that's the same for the  existing `camera` and `switch` so I kept it that way. Gonna change it 🏃🏻💨

joostlek commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,52 @@
+        self.robot = robot
+        self._attr_name = f"{self.robot.name} Dismiss Alert"
+        self._attr_unique_id = self.robot.serial
+        self._attr_entity_category = EntityCategory.CONFIG
```
This one is static and can be set outside of the `__iniit__`

maxcanna committed:
fixup! Add neato dismiss alert button

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

maxcanna committed:
fixup! Add neato dismiss alert button

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

maxcanna committed:
fixup! Add neato dismiss alert button

maxcanna committed:
fixup! Add neato dismiss alert button

gjohansson-ST commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,55 @@
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Neato button from config entry."""
+    dev = []
+
+    for robot in hass.data[NEATO_ROBOTS]:
+        dev.append(NeatoDismissAlertButton(robot))
+
+    if not dev:
+        return
+
+    _LOGGER.debug("Adding vacuum buttons %s", dev)
+    async_add_entities(dev, True)
```
```suggestion
    entities = [
        NeatoDismissAlertButton(robot)
        for robot in hass.data[NEATO_ROBOTS]
    ]

    _LOGGER.debug("Adding vacuum buttons %s", entities)
    async_add_entities(entities, True)
```

gjohansson-ST replied:
Not sure that debugging is actually providing something valuable?

maxcanna replied:
I removed the debug, I kept it since I thought I is good practice to have it watching around

gjohansson-ST commented on homeassistant/components/neato/button.py:
```diff
@@ -0,0 +1,55 @@
+    ) -> None:
+        """Initialize a dismiss_alert Neato button entity."""
+        self.robot = robot
+        self._attr_name = f"{self.robot.name} Dismiss Alert"
+        self._attr_unique_id = f"{self.robot.serial}_dismiss_alert"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(NEATO_DOMAIN, self.robot.serial)}
+        )
```
```suggestion
        self._attr_name = f"{robot.name} Dismiss Alert"
        self._attr_unique_id = f"{robot.serial}_dismiss_alert"
        self._attr_device_info = DeviceInfo(
            identifiers={(NEATO_DOMAIN, robot.serial)}
        )
```

gjohansson-ST reviewed (dismissed):
Tiny remarks left and then we're good

maxcanna committed:
fixup! Add neato dismiss alert button

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

maxcanna committed:
fixup! Add neato dismiss alert button

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

maxcanna committed:
fixup! Add neato dismiss alert button

maxcanna committed:
fixup! Add neato dismiss alert button

joostlek reviewed (approved):
Thanks!
